x-pack/plugins/enterprise_search/server/utils/create_ml_inference_pipeline.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IngestGetPipelineResponse, IngestPipeline } from '@elastic/elasticsearch/lib/api/types';
+import { ElasticsearchClient } from '@kbn/core/server';
+
+import { ErrorCode } from '../../common/types/error_codes';
+
+import { formatMlPipelineBody } from './create_pipeline_definitions';
+
+/**
+ * Details of a created pipeline.
+ */
+export interface CreatedPipeline {
+  id: string;
+  created?: boolean;
+  addedToParentPipeline?: boolean;
+}
+
+/**
+ * Creates a Machine Learning Inference pipeline with the given settings, if it doesn't exist yet,
+ * then references it in the "parent" ML Inference pipeline that is associated with the index.
+ * @param indexName name of the index this pipeline corresponds to.
+ * @param pipelineName pipeline name set by the user.
+ * @param modelId model ID selected by the user.
+ * @param sourceField The document field that model will read.
+ * @param destinationField The document field that the model will write to.
+ * @param esClient the Elasticsearch Client to use when retrieving pipeline and model details.
+ */
+export const createAndReferenceMlInferencePipeline = async (
+  indexName: string,
+  pipelineName: string,
+  modelId: string,
+  sourceField: string,
+  destinationField: string,
+  esClient: ElasticsearchClient
+): Promise<CreatedPipeline> => {
+  const createPipelineResult = await createMlInferencePipeline(
+    pipelineName,
+    modelId,
+    sourceField,
+    destinationField || modelId,
+    esClient
+  );
+
+  const addSubPipelineResult = await addSubPipelineToIndexSpecificMlPipeline(
```

Comment:
Looking at this - it may be best to check for the existence of the processor in the `ml-inference` pipeline definition before calling the `createMlInferencePipeline` method -- this way, it will catch this early - _unless_ the intent is to create the pipeline regardless of whether or not it exists in the parent...

Theoretically, it should never exist (as we should check for uniqueness of the name) - but, just in case...

Reply:
@markjhoy Not sure what the benefit of that extra step would be... We check the existence of the `ml-inference-${userDefinedPipelineName}` pipeline as the first step in `createMlInferencePipeline()`. Given that the `@ml-inference` parent pipeline's processor references should be in sync with the corresponding inference pipeline instances (unless the data is corrupted), the preemptive check for the reference wouldn't provide us with extra information. Am I missing something?

Reply:
Gotcha - that makes sense...